File: day7.dart
import 'dart:io';

const start = 'S';
const splitter = '^';

void main() {
    var lines = File('day7.input').readAsLinesSync();
    part1(lines);
    part2(lines);
}

int sum(a, b) { return a+b; }

void part1(List lines) {
    Set beams = { lines.first.indexOf(start) };
    int count = 0;
    for (final line in lines.sublist(1)) {
        Set splitters = splitter.allMatches(line).map((m) => m.start).toSet();
        Set beamsToSplit = beams.intersection(splitters);
        if (beamsToSplit.isNotEmpty) {
            count += beamsToSplit.length;
            Set splitBeams = {};
            for (final beam in beamsToSplit) {
                splitBeams.add(beam-1);
                splitBeams.add(beam+1);
            }
            beams = beams.difference(beamsToSplit);
            beams = beams.union(splitBeams);
        }
    }
    print(count);
}

void part2(List lines) {
    List beams = List.filled(lines.first.length, 0);
    beams[lines.first.indexOf(start)] = 1;
    for (final line in lines.sublist(1)) {
        for (final i in splitter.allMatches(line).map((m) => m.start)) {
            int n = beams[i];
            if (n > 0) {
                beams[i-1] += n;
                beams[i+1] += n;
                beams[i] = 0;
            }
        }
    }
    print(beams.reduce(sum));
}

File: day7.jl
# day7.jl - Advent of Code 2025

const start, splitter = 'S', '^'

function part1((head, tail...))
    beams = Set(findall(start, head))
    count = 0
    for line in tail
        splitters = Set(findall(splitter, line))
        tosplit = intersect(beams, splitters)
        count += length(tosplit)
        beams = setdiff(beams, tosplit)
        beams = union(beams, vcat(tosplit.-1, tosplit.+1))
    end
    count
end

function part2((head, tail...))
    beams = [c == start ? 1 : 0 for c in head]
    for line in tail
        for i in findall(splitter, line)
            n = beams[i]
            beams[i-1] += n
            beams[i+1] += n
            beams[i] = 0
        end
    end
    sum(beams)
end

lines = "day7.input" |> readlines
lines |> part1 |> println
lines |> part2 |> println
